algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala

```diff
@@ -812,18 +827,44 @@ case class HeavyHitter[K](item: K, count: Long) extends java.io.Serializable
 
 object HeavyHitter {
 
-  def ordering[K: Ordering]: Ordering[HeavyHitter[K]] = Ordering.by { hh: HeavyHitter[K] => (hh.count, hh.item) }
+  import scala.reflect._
+
+  // Note that we can't define the ordering on `count` only.  We use a set to track heavy hitters, where equality means
+  // two elements are identical.  If we were to use `count` only, we would label different items in heavy hitters as
+  // identical simply because they have the same count (and such false duplicates may happen easily in practice).  For
+  // this reason we need to add another dimension to the ordering to prevent such false duplicates.  For this second
+  // dimension we use an item's default Java hash code because we do not want to introduce an `Ordering` constraint.
+  def ordering[K: ClassTag]: Ordering[HeavyHitter[K]] = {
```

Comment:
I'd be more inclined to say "Damn Java Array!" but otherwise I totally agree. ;-)

> Equiv and scala.util.hashing.Hashing should be what we want here, but they are so broken and impoverished I guess we should think about it.

I haven't used either of those yet, but I understand that you would not recommend them.  (Did you mean "I guess we should /not/ think about it"?)

> I really don't like using ClassTag in general, and I especially don't like it here when it is in fact a round-about way to get at what we really want, which is a hashcode.
> 
> So, I guess the least ugly solution is to keep the Ordering[K] bound on cases with the TopK heavy hitters. What do you think?

I'm torn.  While I don't like `ClassTag` shenanigans that much either -- particularly when they are just a workaround for a different issue (read: Array's messed up equality) -- there's one reason I'd prefer `ClassTag` over `Ordering`:  Not having the `Ordering[K]` bound clarifies the contract.  There is no reason the user should have to provide an `Ordering` -- we only use an Ordering internally to benefit from better update performance of a `SortedSet`, and as you pointed out this is mostly because of the TopK use case.  We do not guarantee any order(ing) at all in the API, which only guarantees a `Set` when retrieving the heavy hitters from a CMS instance.  In other words, requiring an `Ordering` kinda leaks internal, implementation-specific information to the user (well, at least in the current scenario where the API does not guarantee a `SortedSet`).
    Another reason would be to keep the code and the look'n'feel as identical as possible between TopPct and TopK.

On the other hand, keeping the Ordering bound for TopK only might be a sneaky transition path to eventually get rid of TopK altogether, though @jnievelt will not like to hear that. ;-)

> So, I guess the least ugly solution is to keep the Ordering[K] bound on cases with the TopK heavy hitters.

There's also a tiny performance benefit of using an `Ordering` internally for TopPct, as we don't have to traverse the full list of heavy hitters every time we perform an update because `dropWhile` can terminate earlier than `filter`:

```scala
// TopPct with Ordering
override def purgeHeavyHitters(cms: CMS[K])(hitters: HeavyHitters[K]): HeavyHitters[K] = {
  val minCount = heavyHittersPct * cms.totalCount
  // with Ordering/SortedSet: HeavyHitters[K](hitters.hhs.dropWhile { _.count < minCount })
  HeavyHitters[K](hitters.hhs.filter { _.count >= minCount })
}
```

Whether that's an important optimization to warrant keeping `Ordering` I can't tell.  Typically, the list of heavy hitters shouldn't be that huge that terminating early is that big of a benefit, for example.